### PR TITLE
Implemented unlock dialog for auto type (bug #197)

### PR DESCRIPTION
**Pull request by sevanteri**
I implemented a dialog for unlocking the first database tab when all databases are locked and the autotype hotkey is pressed. 
Link to the issue: https://www.keepassx.org/dev/issues/197

### CODE REVIEW

**Comment by sevanteri** [comment]
Travis seems to fail because of the GUI changes.

File: src/gui/DatabaseTabWidget.cpp [hunk 1]
```diff
@@ -713,5 +714,9 @@ void DatabaseTabWidget::performGlobalAutoType()
         }
     }
 
-    autoType()->performGlobalAutoType(unlockedDatabases);
+    if (unlockedDatabases.count() == 0) {
+        indexDatabaseManagerStruct(0).dbWidget->showUnlockDialog();
```
Comment (debfx) [inline src/gui/DatabaseTabWidget.cpp]: Will be very unhappy if there is no database tab.
You'll want to iterate over the databases and show the dialog for the first database that is actually locked.

File: src/gui/UnlockDatabaseDialog.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    explicit UnlockDatabaseDialog(QWidget *parent = Q_NULLPTR);
+    void setDBFilename(const QString& filename);
+    void clearForms();
+    Database* database();
+
+Q_SIGNALS:
+    void unlockDone(bool);
+
+public Q_SLOTS:
+    void Done(bool r);
```
Comment (debfx) [inline src/gui/UnlockDatabaseDialog.h]: Please rename as methods start with a lower case letter and done() already exists in QDialog.

File: src/gui/UnlockDatabaseDialog.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+private:
+    UnlockDatabaseWidget* const m_view;
+    DatabaseManagerStruct m_dbStruct;
```
Comment (debfx) [inline src/gui/UnlockDatabaseDialog.h]: Seems unused?

File: src/gui/UnlockDatabaseDialog.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#ifndef KEEPASSX_AUTOTYPEUNLOCKDIALOG_H
+#define KEEPASSX_AUTOTYPEUNLOCKDIALOG_H
+
+#include <QAbstractItemModel>
+#include <QObject>
+#include <QDialog>
+
+#include <gui/DatabaseTabWidget.h>
```
Comment (debfx) [inline src/gui/UnlockDatabaseDialog.h]: #include "gui/DatabaseTabWidget.h"

File: src/gui/UnlockDatabaseDialog.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#ifndef KEEPASSX_AUTOTYPEUNLOCKDIALOG_H
+#define KEEPASSX_AUTOTYPEUNLOCKDIALOG_H
+
+#include <QAbstractItemModel>
```
Comment (debfx) [inline src/gui/UnlockDatabaseDialog.h]: Unused include.

File: src/gui/DatabaseWidget.cpp [hunk 6]
```diff
@@ -643,7 +648,11 @@ void DatabaseWidget::unlockDatabase(bool accepted)
         return;
     }
 
-    replaceDatabase(static_cast<DatabaseOpenWidget*>(sender())->database());
+    // check if sender() is either the unlock widget or the unlock dialog
+    Database *db = sender() == m_unlockDatabaseDialog ? m_unlockDatabaseDialog->database() :
```
Comment (debfx) [inline src/gui/DatabaseWidget.cpp]: Please use if (), else if(), else for readability.

**Comment by debfx** [comment]
UnlockDatabaseDialog should be closed once the database has been unlocked or closed from the main window.

**Comment by sevanteri** [comment]
Right'o.

Currently auto type is performed on all of the unlocked databases. If all of the databases are locked, then a dialog is shown for the first tab.

And I just realized there is no point iterating the tabs if all of them are locked. Useless for loop there...

**Pull request closed without merging**